apkid/rules/elf/obfuscators.yara

```diff
@@ -130,7 +131,87 @@ rule ollvm_v6_0 : obfuscator
     $based_on      = "(based on Obfuscator-LLVM 6.0."
 
   condition:
-    all of them and not ollvm_v6_0_strenc
+    is_elf and
+    all of them and
+    not ollvm_v6_0_strenc
+}
+
+rule ollvm_v9_strenc : obfuscator
+{
+  meta:
+    description = "Obfuscator-LLVM version 9.x (string encryption)"
+    sample      = "2314ec0053d829d424a82f702188fcb525cefce4feeef096f0855339b897a5d1"
+    author      = "Eduardo Novella"
+
+  strings:
+    $clang_version = /clang version \d\.\d\.\d /
+    $strenc        = /\.datadiv_decode[0-9]{18,20}/
+    $ollvm         = "(based on Obfuscator-LLVM 9."
+
+  condition:
+    is_elf and all of them
+}
+
+rule ollvm_tll : obfuscator
+{
+  meta:
+    description = "Obfuscator-LLVM TLL (string encryption)"
+    url         = "https://github.com/yazhiwang/ollvm-tll.git"
+    sample      = "1f010330e9ac90f00d11aa37fdca25c437ad6f4b1302f6d7aa48b91ef22cc107"
+    author      = "Eduardo Novella"
+
+  strings:
+    /**
+      .datadiv_decode7760209850571766755
+      Android clang version 5.0.300080  (based on LLVM 5.0.300080)
+      clang version 6.0.0 (tags/RELEASE_600/final) (https://github.com/yazhiwang/ollvm-tll.git a38559e4c13359073102793c0a734bb1add3d5ff)
+    */
+    $clang_version = /clang version \d\.\d\.\d \(tags\/RELEASE\_\d{3}\/final\)/
+    $strenc        = /\.datadiv_decode[0-9]{18,20}/
+    $url           = "https://github.com/yazhiwang/ollvm-tll.git"
+
+  condition:
+    is_elf and all of them
+}
+
+rule ollvm_armariris : obfuscator
+{
+  meta:
+    description = "Armariris Obfuscator-LLVM (string encryption)"
+    url         = "https://github.com/GoSSIP-SJTU/Armariris"
+    sample      = "d22c2f53bab6fa2ab7bdb4f7acabb419e3ee3163bb758c4f7a013d07a8b09e12" // aka malware Joker
+    author      = "Eduardo Novella"
+
+  strings:
+    // clang version 3.9.1 (tags/RELEASE_391/final)
+    // clang version 5.0.1 (tags/RELEASE_501/final)
+    // .datadiv_decode14660921177804423408
+    $clang_version = /clang version \d\.\d\.\d \(tags\/RELEASE\_\d{3}\/final\)/
+    $strenc        = /\.datadiv_decode[0-9]{18,20}/
+
+  condition:
+    is_elf and
+    all of them and
+    not ollvm_tll and
+    for any i in (0..elf.symtab_entries): (elf.symtab[i].name contains ".datadiv_decode")
+}
+
+rule ollvm_strenc : obfuscator
+{
+  meta:
+    description = "Obfuscator-LLVM alike (string encryption)"
+    sample      = "73f34f7dd5f5c2eff33fc48371c850a2a3ff0355a2bfa014467478ccb30309e3"
+    author      = "Eduardo Novella"
+
+  strings:
+    $strenc = /\.datadiv_decode[0-9]{18,20}/
```

Comment:
This one rule shows up so many times, I think it's worth it to make a private internal rule for it. Why does this only look for the string and not the symtab_entries? If you _can_ look in the symtab entries, that predicate should also go in the private internal rule.

Reply:
Already explained in this post to Tim. The binary is packed with UPX, and symbols are not defined as such, instead we have it as string.

Reply:
So, if this is the case - can we just add a rule to detect this obfuscation + UPX? Instead of just throwing proper detection out the window because one sample used UPX as well.

Something like..

if has symtab_entries then X
if !symtab_entires but symtab_entry string AND UPX, then X that's been UPX packed?

This would clearly separate the two detections, right?

Reply:
Makes sense, but i got no idea where i found this sample. Probably one of these? Let me check now....

```sh
2314ec0053d829d424a82f702188fcb525cefce4feeef096f0855339b897a5d1.apk
45267269b56e889fd653259656c7f7422732180747fa3f240bec7ab98b089f99.apk
65153c8b4e990abf8654291fc5f6f2ddf3e05200b52f59ff6e37f1ae871d0b2f.apk
bb7b0408aa8ce97280c8806bb640f7e0b7c11a0c8fef4186d529b1a4df199fa0.apk
d7bf9517533446cecb76b29711d8e3fb14014c21ab3f2ffb5f13d3cfba2c4158.apk
e8f1a2b0c4b848bc485b60880bd1a97cb66d72403f6a504d2a6ec23662328d9b.apk
e942d9844da8e457ef05b28087f1b33d60f965158b645be78b5162fe1c886bcd.apk
```

Reply:
Much appreciated! I don't _think_ it would be too complicated? Though it'd be similar to how UPX modified/known/malicious detection is done.

Reply:
What's wrong with that? Or better said, what could improve on this?
```sh
[*] d7bf9517533446cecb76b29711d8e3fb14014c21ab3f2ffb5f13d3cfba2c4158.apk!assets/libmgRun_04.22.22_00.so
 |-> obfuscator : Obfuscator-LLVM version unknown (string encryption)
 |-> packer : UPX (unknown, modified)
```

Reply:
I was under the impression, based on your comments above, that neither of these are unknown, no?


Basically... Why are we /adding/ a detection rule, explicitly for something to be called unknown. If it's your known version, you know how to detect it and it's just UPX packed, wouldn't this be more interesting to you/others to know that... hey - this is UPX pack version X?

So if/when a new one rolls out that /is not/ the same (thus unknown) we can have unknown catches?

This coincides with the conversation above about - is this "unknown" or what? So maybe it depends on what you do there...